felix/fv/pre_dnat_test.go

```diff
@@ -200,6 +206,87 @@ var _ = infrastructure.DatastoreDescribe("pre-dnat with initialized Felix, 2 wor
 				})
 
 				Context("with pre-DNAT policy to open pinhole to 8055", func() {
+					// Exec iptables-save, or iptables-save -t [table], if table is not "".
+					dumpIPTables := func(table string) []string {
+						execArgs := []string{"iptables-save"}
+						if table != "" {
+							execArgs = append(execArgs, "-t", table)
+						}
+						out, err := tc.Felixes[0].ExecOutput(execArgs...)
+						Expect(err).NotTo(HaveOccurred())
+						lines := strings.Split(out, "\n")
+						Expect(len(lines)).NotTo(Equal(1), "couldn't split on newline")
+						var grep []string
+						for _, line := range lines {
+							if strings.Contains(line, "cali-") && strings.Contains(line, "-A ") {
+								grep = append(grep, line)
+							}
+						}
+						return grep
+					}
+					dumpMangleTable := func() []string { return dumpIPTables("mangle") }
+					dumpFilterTable := func() []string { return dumpIPTables("filter") }
+
+					// Returns nil when the value of f() hasn't changed for the duration settledPeriod.
+					// Returns error if f() never settles up to the timeout.
+					// Immediately returns error if timeout is lower than settledPeriod.
+					waitToSettle := func(f func() int, settledPeriod, timeout time.Duration) error {
+						if timeout < settledPeriod {
+							return errors.New("Programmer error: timeout is less than settledPeriod")
+						}
+						var last int
+						var settledTimer, timeoutTimer *time.Timer
+						timeoutTimer = time.NewTimer(timeout)
+						for {
+							cur := f()
+
+							if settledTimer != nil {
+								if last != cur {
+									if !settledTimer.Stop() {
+										<-settledTimer.C
+									}
+									settledTimer = nil
+								} else {
+									select {
+									case <-settledTimer.C:
+										// We've been settled long enough, return
+										return nil
+									default:
+										// Still settled, but not for long enough. Carry on...
+									}
+								}
+							} else {
+								if last == cur {
+									settledTimer = time.NewTimer(settledPeriod)
+								}
+							}
+							select {
+							case <-timeoutTimer.C:
```

Comment:
Using a sleep and a timeout channel is slightly mixing paradigms.  While you''re asleep you're not checking the timeout channel, which could be a bug (or not matter at all).  In this case, it means that, if you just miss a timeout, it'll be >1s before you check the timeout again and you'll do an extra poll of `f()`; not clear if that was intended (it probably doesn't matter for this use case).

Best to jump one way or the other to make your intent clear.  Either:
* Use Sleep() and record the start time and do `if time.Since(startTime) > timeoutPeriod { ... timeout ...}`
* Use channels for the sleep.  `case <-time.After(delay): ...` so that the sleep can be interrupted.

Reply:
If you also wanted the settled timer to interrupt the sleep (not clear that you do, since you probably want to run f() one more time before declaring it settled), then... A powerful technique is to mask channels in your main select block.  To mask a channel, set it to nil; that prevents its case from firing.  So you could do:
```
var settledC chan <-time.Time

for {
  select {
  case <- time.After(time.Second):
    cur := f()
    if cur == last {
      if settledC == nil { 
        settledC = time.After(...)
      }
    } else {
      settledC  = nil
    }
  case <- settledC:
    // done!
  case <-timeoutC:
    // failed
  }
}
```